core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java

```diff
@@ -11,53 +11,240 @@
+import alluxio.conf.PropertyKey;
+import alluxio.resource.LockResource;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import org.apache.zookeeper.server.ByteBufferInputStream;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
 import java.nio.channels.ReadableByteChannel;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
+import javax.annotation.concurrent.GuardedBy;
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
- * A class to manage cached pages. This class will
- * 1. Ensure thread-safety
- * 2. Bookkeep Cache Replacement Alg
+ * A class to manage cached pages. This class coordinates different components to respond for
+ * thread-safety and operate cache replacement policies.
  *
+ * Lock hierarchy in this class: All operations must follow this order to operate on pages:
+ * <ul>
+ * <li>1. Acquire page lock</li>
+ * <li>2. Acquire metastore lock mMetaLock</li>
+ * <li>3. Release metastore lock mMetaLock</li>
+ * <li>4. Release page lock</li>
+ * </ul>
  */
 @ThreadSafe
 public class LocalCacheManager implements CacheManager {
   private static final Logger LOG = LoggerFactory.getLogger(LocalCacheManager.class);
 
-  private final CacheEvictor mEvictor = CacheEvictor.create();
-  private final PageStore mPageStore = PageStore.create();
-  private final MetaStore mMetaStore = new MetaStore();
+  private static final int LOCK_SIZE = 1024;
+  private final int mPageSize;
+  private final long mCacheSize;
+  private final CacheEvictor mEvictor;
+  /** A readwrite lock pool to guard individual pages based on striping. */
+  private final ReadWriteLock[] mPageLocks = new ReentrantReadWriteLock[LOCK_SIZE];
+  private final PageStore mPageStore;
+  /** A readwrite lock to guard metadata operations. */
+  private final ReadWriteLock mMetaLock = new ReentrantReadWriteLock();
+  @GuardedBy("mMetaLock")
+  private final MetaStore mMetaStore;
+  private final FileSystemContext mFsContext;
 
-  public LocalCacheManager() {
+  /**
+   * @param fsContext filesystem context
+   */
+  public LocalCacheManager(FileSystemContext fsContext) {
+    this(fsContext, new MetaStore(), PageStore.create(), CacheEvictor.create());
   }
 
-  @Override
-  public int put(long fileId, long pageId, byte[] page) throws IOException {
-    mMetaStore.addPage(pageId);
-    mPageStore.put(fileId, pageId, page);
-    return 0;
+  /**
+   * @param fsContext filesystem context
+   */
+  @VisibleForTesting
+  LocalCacheManager(FileSystemContext fsContext, MetaStore metaStore,
+                    PageStore pageStore, CacheEvictor evictor) {
+    mFsContext = fsContext;
+    mMetaStore = metaStore;
+    mPageStore = pageStore;
+    mEvictor = evictor;
+    mPageSize = (int) mFsContext.getClusterConf().getBytes(PropertyKey.USER_CLIENT_CACHE_PAGE_SIZE);
```

Comment:
Could you update the page size constant used in `LocalCacheFileInStream`?

Reply:
Updated.